# Gemini programming code responses test for diabetes dataset

### Prompt 1

In [2]:
#import pandas as pd; df = pd.read_csv('diabetesdatasets/diabetic_data.csv', nrows=5); print(df.columns.tolist()); print(df.head())

In [3]:
import pandas as pd
df = pd.read_csv('diabetesdatasets/diabetic_data.csv'); print('Race Distribution:\n', df['race'].value_counts()); print('\nGender Distribution:\n', df['gender'].value_counts()); print('\nAge Distribution:\n', df['age'].value_counts())

Race Distribution:
 race
Caucasian          76099
AfricanAmerican    19210
?                   2273
Hispanic            2037
Other               1506
Asian                641
Name: count, dtype: int64

Gender Distribution:
 gender
Female             54708
Male               47055
Unknown/Invalid        3
Name: count, dtype: int64

Age Distribution:
 age
[70-80)     26068
[60-70)     22483
[50-60)     17256
[80-90)     17197
[40-50)      9685
[30-40)      3775
[90-100)     2793
[20-30)      1657
[10-20)       691
[0-10)        161
Name: count, dtype: int64


### output
output same as what llm said what the output is

## Prompt 2

In [4]:
import pandas as pd
df = pd.read_csv('diabetesdatasets/diabetic_data.csv'); print('Readmission rates by Race:\n'); print(pd.crosstab(df['race'], df['readmitted'], normalize='index') * 100)

Readmission rates by Race:

readmitted             <30        >30         NO
race                                            
?                 8.271007  23.669160  68.059833
AfricanAmerican  11.218116  34.534097  54.247788
Asian            10.140406  25.117005  64.742590
Caucasian        11.290556  35.643044  53.066400
Hispanic         10.407462  31.516937  58.075601
Other             9.628154  29.614874  60.756972


In [5]:
import pandas as pd
df = pd.read_csv('diabetesdatasets/diabetic_data.csv'); print('\nReadmission rates by Gender:\n'); print(pd.crosstab(df['gender'], df['readmitted'], normalize='index') * 100)


Readmission rates by Gender:

readmitted             <30        >30          NO
gender                                           
Female           11.245156  35.676683   53.078160
Male             11.061524  34.060142   54.878334
Unknown/Invalid   0.000000   0.000000  100.000000


In [6]:
import pandas as pd
df = pd.read_csv('diabetesdatasets/diabetic_data.csv'); print('\nReadmission rates by Age:\n'); print(pd.crosstab(df['age'], df['readmitted'], normalize='index') * 100)


Readmission rates by Age:

readmitted        <30        >30         NO
age                                        
[0-10)       1.863354  16.149068  81.987578
[10-20)      5.788712  32.416787  61.794501
[20-30)     14.242607  30.778515  54.978877
[30-40)     11.231788  31.443709  57.324503
[40-50)     10.604027  33.846154  55.549819
[50-60)      9.666203  34.289522  56.044274
[60-70)     11.128408  35.124316  53.747276
[70-80)     11.773055  36.347246  51.879699
[80-90)     12.083503  36.186544  51.729953
[90-100)    11.099177  28.929467  59.971357


# comment
same output as what llms said output would be

## Prompt 3

In [7]:
import pandas as pd
df = pd.read_csv('diabetesdatasets/diabetic_data.csv'); print('Race Distribution (Counts):\n', df['race'].value_counts()); print('\nRace Distribution (Percentage):\n', df['race'].value_counts(normalize=True) * 100)

Race Distribution (Counts):
 race
Caucasian          76099
AfricanAmerican    19210
?                   2273
Hispanic            2037
Other               1506
Asian                641
Name: count, dtype: int64

Race Distribution (Percentage):
 race
Caucasian          74.778413
AfricanAmerican    18.876639
?                   2.233555
Hispanic            2.001651
Other               1.479866
Asian               0.629876
Name: proportion, dtype: float64


In [9]:
import pandas as pd
df = pd.read_csv('diabetesdatasets/diabetic_data.csv'); print('\nAge Distribution (Counts):\n', df['age'].value_counts()); print('\nAge Distribution (Percentage):\n', df['age'].value_counts(normalize=True) * 100)


Age Distribution (Counts):
 age
[70-80)     26068
[60-70)     22483
[50-60)     17256
[80-90)     17197
[40-50)      9685
[30-40)      3775
[90-100)     2793
[20-30)      1657
[10-20)       691
[0-10)        161
Name: count, dtype: int64

Age Distribution (Percentage):
 age
[70-80)     25.615628
[60-70)     22.092840
[50-60)     16.956547
[80-90)     16.898571
[40-50)      9.516931
[30-40)      3.709490
[90-100)     2.744532
[20-30)      1.628245
[10-20)      0.679009
[0-10)       0.158206
Name: proportion, dtype: float64


### comment
same output as what llm said it would be

In [11]:
#old code provided by LLM
# --- Conceptual Code Snippets ---
# Assume X_train, y_train are your training features and target
# And 'race' or 'age' is one of the columns in X_train or affects y_train's balance
# For simplicity, let's imagine we're rebalancing based on the target variable 'y_train'
# which itself might be imbalanced, and this imbalance could be correlated with 'race' or 'age'.

# 0. Setup (Illustrative - you would load your data properly)
# import pandas as pd
# from sklearn.model_selection import train_test_split
# from collections import Counter

# # Load your data (replace with actual loading)
# # df = pd.read_csv('diabetesdatasets/diabetic_data.csv')
# # # Preprocess: e.g., handle '?' in race, encode categoricals, define X and y
# # # For illustration, assume 'y' is a binary target derived from 'readmitted'
# # y = df['readmitted'].apply(lambda x: 1 if x == '<30' else 0)
# # X = df.drop('readmitted', axis=1)
# # # ... further preprocessing for X (encoding, scaling, feature selection) ...
# # X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42, stratify=y)


# 1. Handling '?' in Race (Example: Mode Imputation)
# X_train['race'] = X_train['race'].replace('?', X_train['race'].mode()[0])
# X_test['race'] = X_test['race'].replace('?', X_train['race'].mode()[0]) # Use mode from train set

# 2. SMOTE (Oversampling)
# from imblearn.over_sampling import SMOTE
# print(f"Original training set shape: {Counter(y_train)}")
# smote = SMOTE(random_state=42, sampling_strategy='auto') # Adjust sampling_strategy as needed
# X_train_smote, y_train_smote = smote.fit_resample(X_train, y_train)
# print(f"Resampled training set shape (SMOTE): {Counter(y_train_smote)}")

# 3. RandomUnderSampler (Undersampling)
# from imblearn.under_sampling import RandomUnderSampler
# print(f"Original training set shape: {Counter(y_train)}")
# rus = RandomUnderSampler(random_state=42, sampling_strategy='auto') # sampling_strategy='majority'
# X_train_rus, y_train_rus = rus.fit_resample(X_train, y_train)
# print(f"Resampled training set shape (RandomUnderSampler): {Counter(y_train_rus)}")

# 4. SMOTETomek (Hybrid)
# from imblearn.combine import SMOTETomek
# print(f"Original training set shape: {Counter(y_train)}")
# smt = SMOTETomek(random_state=42, smote=SMOTE(sampling_strategy='auto'), tomek=TomekLinks(sampling_strategy='all'))
# X_train_smt, y_train_smt = smt.fit_resample(X_train, y_train)
# print(f"Resampled training set shape (SMOTETomek): {Counter(y_train_smt)}")

# 5. Cost-Sensitive Learning (Example with RandomForestClassifier)
# from sklearn.ensemble import RandomForestClassifier
# # Define class weights - e.g., if class 0 is majority and class 1 is minority
# # weights = {0: 1, 1: 10} # Give 10x more weight to misclassifying class 1
# # model = RandomForestClassifier(class_weight=weights, random_state=42)
# # model.fit(X_train, y_train) # Train on original (but possibly preprocessed) data

#New code
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from collections import Counter

from imblearn.over_sampling import SMOTE
from imblearn.under_sampling import RandomUnderSampler
from imblearn.combine import SMOTETomek
from sklearn.ensemble import RandomForestClassifier

# 1. Load the data
df = pd.read_csv("diabetesdatasets/diabetic_data.csv")

# 2. Define your binary target: readmitted within 30 days
df["readmitted"] = df["readmitted"].apply(lambda x: 1 if x == "<30" else 0)
y = df["readmitted"]
X = df.drop(columns=["readmitted"])

# 3. Impute missing race entries (‘?’) with the mode of the training set
mode_race = X["race"].mode()[0]
X["race"] = X["race"].replace("?", mode_race)

# 4. Encode all categorical columns as integer codes
for col in X.select_dtypes(include="object"):
    X[col] = X[col].astype("category").cat.codes

# 5. Split into train and test (stratified on y)
X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.30,
    random_state=42,
    stratify=y
)

print("Original training set shape:", Counter(y_train))

# 6. SMOTE oversampling
smote = SMOTE(random_state=42)
X_train_smote, y_train_smote = smote.fit_resample(X_train, y_train)
print("After SMOTE:", Counter(y_train_smote))

# 7. Random undersampling
rus = RandomUnderSampler(random_state=42)
X_train_rus, y_train_rus = rus.fit_resample(X_train, y_train)
print("After RandomUnderSampler:", Counter(y_train_rus))

# 8. SMOTE + Tomek links (hybrid)
smt = SMOTETomek(random_state=42)
X_train_smt, y_train_smt = smt.fit_resample(X_train, y_train)
print("After SMOTETomek:", Counter(y_train_smt))

# 9. Cost-sensitive learning with RandomForest
#    using balanced class weights
rf = RandomForestClassifier(class_weight="balanced", random_state=42)
rf.fit(X_train, y_train)
print("Trained RandomForest with class_weight='balanced'")

Original training set shape: Counter({0: 63286, 1: 7950})
After SMOTE: Counter({0: 63286, 1: 63286})
After RandomUnderSampler: Counter({0: 7950, 1: 7950})
After SMOTETomek: Counter({0: 55505, 1: 55505})
Trained RandomForest with class_weight='balanced'


##  output
original training set was heavily imbalanced, with about 63,286 negative cases versus 7,950 positives. Applying SMOTE generated synthetic positives until both classes had 63,286 samples, perfectly balancing the data but at the cost of a very large training set. Random undersampling threw away excess negatives to match the 7,950 positives, giving equal class sizes but much less data overall. The hybrid SMOTETomek approach first oversampled and then cleaned border points, yielding 55,505 of each class, a middle ground between data volume and balance. Finally, you trained a RandomForest using class_weight='balanced', which tells the model itself to pay extra attention to the minority class without changing the data.